### PR TITLE
Feat/Add protobuf serialization for containers (std::vector).

### PR DESCRIPTION
**Pull request by lhruby**
# Description
* add protobuf serialization for containers. The first implementation adds serialization for `std::vector<T>`. Note:Primitive type and non-primitive types are treated differently by protobuf.
* add test for deserialization into non-empty object

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] If this is a new component I have added examples.
- [x] I updated the README and related documentation.

### CODE REVIEW

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// Vector (std::vector <-> google::protobuf::RepeatedField)
+//=================================================================================================
+template <typename T, typename ProtoT>
+auto toProto(google::protobuf::RepeatedField<ProtoT>& proto_repeated_field, const std::vector<T>& vec)
+    -> void {
+  proto_repeated_field.Clear();  // Ensure that the repeated field is empty before adding elements.
+  proto_repeated_field.Reserve(static_cast<int>(vec.size()));
+  for (const auto& value : vec) {
+    proto_repeated_field.Add(value);
+  }
```
Comment (filippobrizzi) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:28]: how does this even compile? T and ProtoT are different, you are not serializing the vec.
```suggestion
  for (const auto& value : vec) {
    auto *proto_value = proto_repeated_field.Add();   
    toProto(*proto_value, value);
  }
```

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+template <typename T, typename ProtoT>
+auto fromProto(const google::protobuf::RepeatedField<ProtoT>& proto_repeated_field, std::vector<T>& vec)
+    -> void {
+  vec.clear();  // Ensure that the vector is empty before adding elements.
+  vec.reserve(static_cast<size_t>(proto_repeated_field.size()));
+  for (const auto& proto_value : proto_repeated_field) {
+    vec.push_back(proto_value);
+  }
+}
```
Comment (filippobrizzi) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:39]:
```suggestion
  vec.resize(static_cast<size_t>(proto_repeated_field.size()));
  std::size_t idx = 0;
  for (const auto& proto_value : proto_repeated_field) {
    fromProto(proto_value, vec[idx++]);
  }
}
```

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h (hunk 1 above)
Comment (lhruby) [reply modules/serdes/include/hephaestus/serdes/protobuf/containers.h:28]: Good catch. I think it worked as I was testing it on int.

**Comment by johannes-graeter** [comment]
@lhruby what shall we do about this one revisit or close?

**Comment by lhruby** [comment]
> @lhruby what shall we do about this one revisit or close?

Revisit. Do you want to take it?

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+template <typename T, typename ProtoT>
+auto toProto(google::protobuf::RepeatedPtrField<ProtoT>& proto_repeated_ptr_field, const std::vector<T>& vec)
+    -> void {
+  proto_repeated_ptr_field.Clear();  // Ensure that the repeated field is empty before adding elements.
+  proto_repeated_ptr_field.Reserve(static_cast<int>(vec.size()));
+  for (const auto& value : vec) {
+    toProto(*(proto_repeated_ptr_field.Add()), value);
+  }
+}
```
Comment (filippobrizzi) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:41]: duplicate of above

**Review by filippobrizzi — APPROVED** [review]

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+// in this file, calls to toProto and fromProto must be prefixed with (heph::)serdes::protobuf::
+//=================================================================================================
+
+namespace heph::serdes::protobuf {
+
+//=================================================================================================
+// Vector (std::vector <-> google::protobuf::RepeatedField)
+//=================================================================================================
+template <typename T, typename ProtoT>
+  requires std::is_arithmetic_v<T>
```
Comment (johannes-graeter) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:25]: either use heph::NumericType or add a proper concept for that.
IMO 
```cpp
template<Arithmetic T, typename ProtoT>
```
is a lot more readable

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+auto toProto(google::protobuf::RepeatedField<ProtoT>& proto_repeated_field, const std::vector<T>& vec)
+    -> void {
+  proto_repeated_field.Clear();  // Ensure that the repeated field is empty before adding elements.
+  proto_repeated_field.Reserve(static_cast<int>(vec.size()));
+  proto_repeated_field.Add(vec.begin(), vec.end());
```
Comment (johannes-graeter) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:30]: Nit
```suggestion
  proto_repeated_field.Add(vec.cbegin(), vec.cend());
```

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    -> void {
+  proto_repeated_ptr_field.Clear();  // Ensure that the repeated field is empty before adding elements.
+  proto_repeated_ptr_field.Reserve(static_cast<int>(vec.size()));
+  for (const auto& value : vec) {
+    toProto(*(proto_repeated_ptr_field.Add()), value);
+  }
+}
+
+template <typename T, typename ProtoT>
+  requires std::is_arithmetic_v<T>
```
Comment (johannes-graeter) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:44]: see above

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+auto fromProto(const google::protobuf::RepeatedField<ProtoT>& proto_repeated_field, std::vector<T>& vec)
+    -> void {
+  vec.clear();  // Ensure that the vector is empty before adding elements.
+  vec.reserve(static_cast<size_t>(proto_repeated_field.size()));
+  std::transform(proto_repeated_field.begin(), proto_repeated_field.end(), std::back_inserter(vec),
```
Comment (johannes-graeter) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:49]:
```suggestion
  std::transform(proto_repeated_field.cbegin(), proto_repeated_field.cend(), std::back_inserter(vec),
```

File: modules/serdes/include/hephaestus/serdes/protobuf/containers.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                 [](const auto& proto_value) { return static_cast<T>(proto_value); });
+}
+
+template <typename T, typename ProtoT>
+auto fromProto(const google::protobuf::RepeatedPtrField<ProtoT>& proto_repeated_ptr_field,
+               std::vector<T>& vec) -> void {
+  vec.clear();  // Ensure that the vector is empty before adding elements.
+  vec.reserve(static_cast<size_t>(proto_repeated_ptr_field.size()));
+  std::transform(proto_repeated_ptr_field.begin(), proto_repeated_ptr_field.end(), std::back_inserter(vec),
```
Comment (johannes-graeter) [inline modules/serdes/include/hephaestus/serdes/protobuf/containers.h:58]:
```suggestion
  std::transform(proto_repeated_ptr_field.cbegin(), proto_repeated_ptr_field.cend(), std::back_inserter(vec),
```

**Review by johannes-graeter — APPROVED** [review]
nice only nits
